File: KSNavigationController/KSNavigationController.m
//
//  KSNavigationController.m
//
//  Copyright © 2016 Alex Gordiyenko. All rights reserved.
//

/*
 The MIT License (MIT)
 
 Copyright (c) 2016 A. Gordiyenko
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */

#import "KSNavigationController.h"
#import <Quartz/Quartz.h>

#pragma mark - Stack

@interface _KSStackItem : NSObject {
@public
    id _value;
    _KSStackItem *_next;
}
@end

@implementation _KSStackItem
@end

/** Not thread-safe */
@interface _KSStack : NSObject {
    _KSStackItem *_head;
}
- (void)push:(id)object;
- (id)pop;
- (void)iterateWithBlock:(void(^)(id object))block;

@property (nonatomic, readonly) id headValue;
@property (nonatomic, readonly) NSUInteger count;

@end

@implementation _KSStack

- (void)push:(id)object {
    _KSStackItem *item = [_KSStackItem new];
    item->_value = object;
    item->_next = _head;
    _head = item;
    ++_count;
}

- (id)pop {
    if (_head == nil) {
        [NSException raise:NSInternalInconsistencyException format:@"Popped an empty stack"];
        return nil;
    }
    
    id retVal = _head->_value;
    _head = _head->_next;
    --_count;
    return retVal;
}

- (id)headValue {
    return _head ? _head->_value : nil;
}

- (void)iterateWithBlock:(void(^)(id object))block {
    if (!block) {
        return;
    }
    
    _KSStackItem *item = _head;
    while (item) {
        block(item->_value);
        item = item->_next;
    }
}

@end


#pragma mark - KSNavigationController

@interface KSNavigationController () {
    _KSStack *_stack;
    NSViewController<KSNavigationControllerCompatible> *_rootViewController;
    NSView *_activeView;
    dispatch_once_t _addRootViewOnceToken;
}

@property (nonatomic, readonly) CATransition *transition;

@end

@implementation KSNavigationController

@synthesize transition = _transition;

#pragma mark - Life Cycle

- (instancetype)initWithRootViewController:(NSViewController<KSNavigationControllerCompatible> *)rootViewController {
    if (!rootViewController) {
        [NSException raise:NSInternalInconsistencyException format:@"`rootViewController` can't be nil"];
        return nil;
    }
    
    self = [super init];
    if (self) {
        _rootViewController = rootViewController;
        _rootViewController.navigationController = self;
        _stack = [_KSStack new];
    }
    return self;
}

- (void)loadView {
    self.view = [NSView new];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.wantsLayer = YES;
}

- (void)viewWillAppear {
    [super viewWillAppear];
    
    dispatch_once(&_addRootViewOnceToken, ^{
        if (_rootViewController) {
            _activeView = _rootViewController.view;
            [self addActiveViewAnimated:NO subtype:nil];
        }
    });
}

#pragma mark - Accessors

- (CATransition *)transition {
    if (!_transition) {
        _transition = [CATransition animation];
        _transition.type = kCATransitionPush;
        self.view.animations = @{@"subviews": _transition};
    }
    
    return _transition;
}

- (NSUInteger)viewControllersCount {
    return _stack.count + 1;
}

- (NSViewController *)topViewController {
    if (_stack.count) {
        return _stack.headValue;
    }
    
    return _rootViewController;
}

- (NSViewController<KSNavigationControllerCompatible> *)rootViewController {
    return _rootViewController;
}

#pragma mark - Public Methods

- (void)pushViewController:(NSViewController<KSNavigationControllerCompatible> *)viewController animated:(BOOL)animated {
    [_activeView removeFromSuperview];
    [_stack push:viewController];
    viewController.navigationController = self;
    _activeView = viewController.view;
    [self addActiveViewAnimated:animated subtype:kCATransitionFromRight];
}

- (NSViewController<KSNavigationControllerCompatible> *)popViewControllerAnimated:(BOOL)animated {
    if (_stack.count == 0) {
        return nil;
    }
    
    [_activeView removeFromSuperview];
    NSViewController<KSNavigationControllerCompatible> *retVal = [_stack pop];
    _activeView = [_stack.headValue view];
    if (!_activeView) {
        _activeView = _rootViewController.view;
    }
    
    [self addActiveViewAnimated:animated subtype:kCATransitionFromLeft];
    return retVal;
}

- (NSArray<NSViewController<KSNavigationControllerCompatible> *> *)viewControllers {
    if (!_rootViewController) {
        return nil;
    }
    
    NSMutableArray *retVal = [[NSMutableArray alloc] initWithCapacity:_stack.count + 1];
    [_stack iterateWithBlock:^(id object) {
        [retVal addObject:object];
    }];
    [retVal addObject:_rootViewController];

    return [retVal copy];
}

- (NSArray<NSViewController<KSNavigationControllerCompatible> *> *)popToRootViewControllerAnimated:(BOOL)animated {
    if (_stack.count == 0) {
        return nil;
    }
    
    NSMutableArray *retVal = [[NSMutableArray alloc] initWithCapacity:_stack.count];
    for (NSUInteger i = 0; i < _stack.count; i++) {
        [retVal addObject:[self popViewControllerAnimated:animated]];
    }
    
    return [retVal copy];
}

#pragma mark - Private Methods

- (void)addActiveViewAnimated:(BOOL)animated subtype:(NSString *)subtype {
    if (animated) {
        self.transition.subtype = subtype;
        [self.view.animator addSubview:_activeView];
    } else {
        [self.view addSubview:_activeView];
    }
}

@end

File: KSNavigationController/Swift/KSNavigationController.swift
//
//  KSNavigationController.m
//
//  Copyright © 2016 Alex Gordiyenko. All rights reserved.
//

/*
 The MIT License (MIT)
 
 Copyright (c) 2016 A. Gordiyenko
 
 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be included in all
 copies or substantial portions of the Software.
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 SOFTWARE.
 */

import Cocoa

// MARK: Stack

class _KSStackItem<T> : NSObject {
    var value: T
    var next: _KSStackItem<T>?
    init(_ value: T) {
        self.value = value
    }
}

class _KSStack<T>: NSObject {
    private var _head: _KSStackItem<T>?
    private var _count: UInt = 0
    var headValue: T? {
        get {
            return self._head?.value
        }
    }
    var count: UInt {
        get {
            return self._count
        }
    }
    
    func push(object: T) -> Void {
        let item = _KSStackItem(object)
        item.next = self._head
        self._head = item
        self._count += 1
    }
    
    func pop() -> T? {
        guard self._head != nil else {
            NSException(name: NSInternalInconsistencyException, reason: "Popped an empty stack", userInfo: nil).raise()
            return nil
        }
        
        let retVal = self._head?.value
        self._head = self._head?.next
        self._count -= 1
        return retVal
    }
    
    func iterate(block: (T) -> (Void)) -> Void {
        var item = self._head
        while true {
            if let item = item {
                block(item.value)
            } else {
                break
            }
            
            item = item?.next
        }
    }

}

// MARK: KSNavigationControllerCompatible

protocol KSNavigationControllerCompatible {
    weak var navigationController: KSNavigationController? {get set}
}

// MARK: KSNavigationController

class KSNavigationController: NSViewController {
    // MARK: Properties
    
    private(set) var rootViewController: NSViewController
    
    var viewControllers: [NSViewController] {
        get {
            var retVal = [NSViewController]()
            self._stack.iterate { (object: NSViewController) -> (Void) in
                retVal.append(object)
            }
            
            retVal.append(self.rootViewController)
            return retVal
        }
    }
    
    var viewControllersCount: UInt {
        get {
            return self._stack.count + 1
        }
    }
    
    var topViewController: NSViewController? {
        get {
            if self._stack.count > 0 {
                return self._stack.headValue;
            }
            
            return self.rootViewController;
        }
    }

    private var _activeView: NSView?
    private var _addRootViewOnceToken: dispatch_once_t = 0
    private var _stack: _KSStack<NSViewController> = _KSStack<NSViewController>()
    private var _transition: CATransition {
        get {
            let transition = CATransition()
            transition.type = kCATransitionPush
            self.view.animations = ["subviews": transition]
            return transition
        }
    }
    
    // MARK: Life Cycle
    
    init?(rootViewController: NSViewController) {
        self.rootViewController = rootViewController
        super.init(nibName: nil, bundle: nil)
        if var rootViewController = rootViewController as? KSNavigationControllerCompatible {
            rootViewController.navigationController = self
        } else {
            NSException(name: NSInternalInconsistencyException, reason: "`rootViewController` doesn't conform to `KSNavigationControllerCompatible`", userInfo: nil).raise()
            return nil
        }
    }
    
    required init?(coder: NSCoder) {
        self.rootViewController = NSViewController()
        super.init(coder: coder)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.view.wantsLayer = true
    }
    
    override func viewWillAppear() {
        super.viewWillAppear()
        dispatch_once(&self._addRootViewOnceToken) {
            self._activeView = self.rootViewController.view
            self.addActiveViewAnimated(false, subtype: nil)
        }
    }
    
    override func loadView() {
        self.view = NSView()
    }
    
    // MARK: Public Methods
    
    func pushViewController(viewController: NSViewController, animated: Bool) {
        self._activeView?.removeFromSuperview()
        self._stack.push(viewController)
        if var viewControllerWithNav = viewController as? KSNavigationControllerCompatible {
            viewControllerWithNav.navigationController = self
        }
        
        self._activeView = viewController.view
        self.addActiveViewAnimated(animated, subtype: kCATransitionFromRight)
    }
    
    func popViewControllerAnimated(animated: Bool) -> NSViewController? {
        if self._stack.count == 0 {
            return nil
        }
        
        self._activeView?.removeFromSuperview()
        let retVal = self._stack.pop()
        self._activeView = self._stack.headValue?.view
        if self._activeView == nil {
            self._activeView = self.rootViewController.view
        }
        
        self.addActiveViewAnimated(animated, subtype: kCATransitionFromLeft)
        return retVal
    }
    
    func popToRootViewControllerAnimated(animated: Bool) -> [NSViewController]? {
        if self._stack.count == 0 {
            return nil;
        }
        
        var retVal = [NSViewController]()
        for _ in 1...self._stack.count {
            if let vc = self.popViewControllerAnimated(animated) {
                retVal.append(vc)
            }
        }
        
        return retVal
    }
    
    // MARK: Private Methods
    
    func addActiveViewAnimated(animated: Bool, subtype: String?) {
        if animated {
            self._transition.subtype = subtype
            self.view.animator().addSubview(self._activeView!)
        } else {
            self.view.addSubview(self._activeView!)
        }
    }
}
